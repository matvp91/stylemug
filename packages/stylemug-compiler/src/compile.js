import fnv1a from 'fnv1a';
import { save as extractorSave } from './extractor';
import { extractShorthands } from './shorthands';

function hash(str) {
  let hash = fnv1a(str).toString(16);
  const match = str.match(/[a-zA-Z]/);
  const prepend = (match ? match.pop() : undefined) || 'x';
  return `${prepend}${hash}`;
}

export function compileSelectors(selectors, children, media) {
  let result = {};

  selectors = extractShorthands(selectors);

  for (let key in selectors) {
    const value = selectors[key];

    switch (typeof value) {
      case 'object':
        // @media
        if (/^@/.test(key)) {
          result = Object.assign(
            result,
            compileSelectors(value, children, key)
          );
        }
        // &:focus
        else {
          const child = key.replace(/&/g, '');
          result = Object.assign(
            result,
            compileSelectors(value, children + child, media)
          );
        }
        continue;

      case 'number':
      case 'string':
        const mediaKey = media || '';
        const className = hash(key + value + children + mediaKey);

        const entry = {
          // Provide a key id to match selectors setting the same key,
          // for example, when for a given key, two classes apply with both
          // backgroundColor as propery, we can filter out the first one by specificity.
          keyId: hash(key + children + mediaKey),
          key,
          value,
          children,
          media,
        };

        extractorSave(className, entry);

        result[className] = entry;
        continue;
    }
  }

  return result;
}

export function compileSchema(schema) {
  const result = {};

  for (let key in schema) {
    if (typeof schema[key] !== 'object') {
      throw new Error('Schema is not an object.');
    }
    result[key] = compileSelectors(schema[key], '');
  }

  return result;
}
