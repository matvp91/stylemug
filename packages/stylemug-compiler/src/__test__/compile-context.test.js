import { compileSchema } from '../compile';

describe('compile', () => {
  it('should assert an invalid sheet', () => {
    const { result, reports } = compileSchema({
      default: {
        color: 'red',
        backgroundColor: 'yellow',
      },
      // The rule below fails.
      foo: 'bar',
    });

    expect(result).toMatchSnapshot();
    expect(reports).toMatchSnapshot();
  });

  it('should assert an invalid rule value', () => {
    const { reports } = compileSchema({
      default: {
        color: true,
      },
    });

    expect(reports).toMatchSnapshot();
  });

  it('should assert invalid pseudo classes', () => {
    const { reports } = compileSchema({
      default: {
        '&:iAmAnInvalidPseudoClass': {
          color: true,
        },
      },
    });

    expect(reports).toMatchSnapshot();
  });

  it('should report multiple errors', () => {
    const { reports } = compileSchema({
      foo: 'bar',
      default: {
        '&:iAmAnInvalidPseudoClass': {
          color: true,
        },
      },
    });

    expect(reports).toMatchSnapshot();
  });
});
