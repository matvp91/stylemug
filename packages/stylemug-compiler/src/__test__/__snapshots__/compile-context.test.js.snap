// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compile should assert an invalid rule value 1`] = `
Array [
  Object {
    "message": "Invalid type for color",
  },
]
`;

exports[`compile should assert an invalid sheet 1`] = `
Object {
  "default": Object {
    "b765dfa68": Object {
      "children": "",
      "key": "backgroundColor",
      "keyId": "bb7c5a6d0",
      "media": undefined,
      "value": "yellow",
    },
    "cb34cd11f": Object {
      "children": "",
      "key": "color",
      "keyId": "c3d7e6258",
      "media": undefined,
      "value": "red",
    },
  },
}
`;

exports[`compile should assert an invalid sheet 2`] = `
Array [
  Object {
    "message": "Classname with key foo is not an object",
  },
]
`;

exports[`compile should assert invalid pseudo classes 1`] = `
Array [
  Object {
    "message": "&:iAmAnInvalidPseudoClass is an invalid CSS pseudo class / element",
  },
  Object {
    "message": "Invalid type for color",
  },
]
`;

exports[`compile should report multiple errors 1`] = `
Array [
  Object {
    "message": "Classname with key foo is not an object",
  },
  Object {
    "message": "&:iAmAnInvalidPseudoClass is an invalid CSS pseudo class / element",
  },
  Object {
    "message": "Invalid type for color",
  },
]
`;
